Codegen bug with delegate param of array of custom struct

repro IDL:
    struct BlittableStruct
    {
        Int32 i32;
    };
    delegate BlittableStruct[] ArrayParams_BlittableStruct_Del(BlittableStruct[] a, ref BlittableStruct[] b, out BlittableStruct[] c);

resulting projection:
    inline auto ArrayParams_BlittableStruct_Del::operator()(array_view<TestComp::BlittableStruct const> a, array_view<TestComp::BlittableStruct> b, com_array<TestComp::BlittableStruct>& c) const
    {
        uint32_t result_impl_size{};
        struct struct_TestComp_BlittableStruct* result{}; // <-- bug
        check_hresult((*(impl::abi_t<ArrayParams_BlittableStruct_Del>**)this)->Invoke(a.size(), get_abi(a), b.size(), put_abi(b), impl::put_size_abi(c), put_abi(c), &result_impl_size, &result));
        return com_array<TestComp::BlittableStruct>{ result, result_impl_size, take_ownership_from_abi };
    }

and compile error:
2>D:\git\cswinrt\TestComp\Generated Files\winrt\TestComp.h(2069,192): error C2664: 'int32_t winrt::impl::abi<winrt::TestComp::ArrayParams_BlittableStruct_Del,void>::type::Invoke(uint32_t,winrt::impl::struct_TestComp_BlittableStruct *,uint32_t,winrt::impl::struct_TestComp_BlittableStruct *,uint32_t *,winrt::impl::struct_TestComp_BlittableStruct **,uint32_t *,winrt::impl::struct_TestComp_BlittableStruct **) noexcept': cannot convert argument 8 from 'winrt::TestComp::ArrayParams_BlittableStruct_Del::()::struct_TestComp_BlittableStruct **' to 'winrt::impl::struct_TestComp_BlittableStruct **' (compiling source file Class.cpp)
2>D:\git\cswinrt\TestComp\Generated Files\winrt\TestComp.h(2069,185): message : Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast (compiling source file Class.cpp)

appears that this:
        struct struct_TestComp_BlittableStruct* result{};
should be this:
        struct winrt::impl::struct_TestComp_BlittableStruct* result{};

--- a/.gitignore
+++ b/.gitignore
@@ -11,3 +11,4 @@ packages
 Debug
 Release
 Generated Files
+obj

--- a/cppwinrt/code_writers.h
+++ b/cppwinrt/code_writers.h
@@ -965,7 +965,7 @@ namespace cppwinrt
         }
     }
 
-    static void write_consume_return_type(writer& w, method_signature const& signature)
+    static void write_consume_return_type(writer& w, method_signature const& signature, bool delegate_types)
     {
         if (!signature.return_signature())
         {
@@ -981,6 +981,7 @@ namespace cppwinrt
         %* %{};)";
 
             w.abi_types = true;
+            w.delegate_types = delegate_types;
 
             w.write(format,
                 signature.return_param_name(),
@@ -988,6 +989,7 @@ namespace cppwinrt
                 signature.return_param_name());
 
             w.abi_types = false;
+            w.delegate_types = false;
         }
         else if (category == param_category::object_type || category == param_category::string_type)
         {
@@ -1077,7 +1079,7 @@ namespace cppwinrt
             bind<write_comma_generic_types>(generics),
             method_name,
             bind<write_consume_params>(signature),
-            bind<write_consume_return_type>(signature),
+            bind<write_consume_return_type>(signature, false),
             type,
             get_abi_name(method),
             bind<write_abi_args>(signature),
@@ -2447,7 +2449,7 @@ struct __declspec(empty_bases) produce_dispatch_to_overridable<T, D, %>
                 type_name,
                 bind_list(", ", generics),
                 bind<write_consume_params>(signature),
-                bind<write_consume_return_type>(signature),
+                bind<write_consume_return_type>(signature, true),
                 type_name,
                 bind_list(", ", generics),
                 bind<write_abi_args>(signature),
@@ -2500,7 +2502,7 @@ struct __declspec(empty_bases) produce_dispatch_to_overridable<T, D, %>
                 type_name,
                 type_name,
                 bind<write_consume_params>(signature),
-                bind<write_consume_return_type>(signature),
+                bind<write_consume_return_type>(signature, true),
                 type_name,
                 bind<write_abi_args>(signature),
                 bind<write_consume_return_statement>(signature));

--- a/cppwinrt/type_writers.h
+++ b/cppwinrt/type_writers.h
@@ -106,6 +106,7 @@ namespace cppwinrt
 
         std::string type_namespace;
         bool abi_types{};
+        bool delegate_types{};
         bool param_names{};
         bool consume_types{};
         bool async_types{};
@@ -275,6 +276,10 @@ namespace cppwinrt
                     else if ((name == "Point" || name == "Size" || name == "Rect") && ns == "Windows.Foundation")
                     {
                         write("@::%", ns, name);
+                    }
+                    else if (delegate_types)
+                    {
+                        write("struct impl::struct_%_%", get_impl_name(ns), name);
                     }
                     else
                     {

--- a/test/test/struct_delegate.cpp
+++ b/test/test/struct_delegate.cpp
@@ -1,0 +1,21 @@
+#include "pch.h"
+#include "winrt/test_component.delegates.h"
+
+using namespace winrt;
+using namespace test_component;
+
+TEST_CASE("struct_delegate")
+{
+    Delegates::StructDelegate d = []
+    {
+        return com_array{ Struct{ L"First", L"Second" }, Struct{ L"Third", L"Fourth" } };
+    };
+
+    auto values = d();
+
+    REQUIRE(values.size() == 2);
+    REQUIRE(values[0].First == L"First");
+    REQUIRE(values[0].Second == L"Second");
+    REQUIRE(values[1].First == L"Third");
+    REQUIRE(values[1].Second == L"Fourth");
+}

--- a/test/test/test.vcxproj
+++ b/test/test/test.vcxproj
@@ -417,6 +417,7 @@
     <ClCompile Include="return_params_abi.cpp" />
     <ClCompile Include="single_threaded_observable_vector.cpp" />
     <ClCompile Include="structs.cpp" />
+    <ClCompile Include="struct_delegate.cpp" />
     <ClCompile Include="tearoff.cpp" />
     <ClCompile Include="thread_pool.cpp" />
     <ClCompile Include="uniform_in_params.cpp" />

--- a/test/test_component/test_component.idl
+++ b/test/test_component/test_component.idl
@@ -11,6 +11,12 @@ namespace Windows.Foundation.Metadata
 
 namespace test_component
 {
+    struct Struct
+    {
+        String First;
+        String Second;
+    };
+
     namespace Delegates
     {
         delegate void AgileDelegate();
@@ -22,6 +28,7 @@ namespace test_component
         delegate String[] ReturnStringArrayDelegate();
         delegate void OutStringArrayDelegate(out String[] value);
         delegate void RefStringArrayDelegate(ref String[] value);
+        delegate test_component.Struct[] StructDelegate();
     }
 
     enum Signed
@@ -37,12 +44,6 @@ namespace test_component
         First = 0,
         Second = 1,
         Third = 2
-    };
-
-    struct Struct
-    {
-        String First;
-        String Second;
     };
 
     runtimeclass Simple

